classes/allocator.py

```diff
@@ -52,64 +40,29 @@ def read_entities(self):
                 elif row["type"] == "startup":
                     self.startups.append(Startup(data=row))
 
-    def add_startups(self):
-        for startup in self.startups:
-            for bin in self.bins:
-                bin.add_startup(startup)
-
-    def calc_capacity(self):
-        for bin in self.bins:
-            bin.calc_capacity(self.judges)
-
-    def work_left(self):
-        return any([bin.work_left() for bin in self.bins])
-
-    def default_bins(self):
-        self.bins = (
-            [ReadsBin(),
-             SatisfiedBin(),
-             FemaleBin(weight=FEMALE_WEIGHT*BIN_DEFAULT_WEIGHT),
-             RoleBin(value="Executive",
-                     weight=ROLE_WEIGHT*BIN_DEFAULT_WEIGHT,
-                     count=2),
-             RoleBin(value="Lawyer",
-                     weight=ROLE_WEIGHT*BIN_DEFAULT_WEIGHT),
-             RoleBin(value="Investor",
-                     weight=ROLE_WEIGHT*BIN_DEFAULT_WEIGHT)] +
-            bin_factory(IndustryBin,
-                        values=[value for value, _ in industry.values],
-                        weight=INDUSTRY_WEIGHT*BIN_DEFAULT_WEIGHT) +
-            bin_factory(HomeProgramBin,
-                        values=[value for value, _ in program.values],
-                        weight=HOME_PROGRAM_WEIGHT*BIN_DEFAULT_WEIGHT))
+    def setup(self):
+        self.heuristic.setup(self.judges, self.startups)
 
     def allocate(self):
-        while self.work_left() and self.judges:
+        while self.heuristic.work_left() and self.judges:
             judge = choice(self.judges)
-            judge.next_action(self.bins)
-            self.register_judge_events(judge)
-            if not judge.has_more_work:
+            self.heuristic.process_judge_events(judge.complete_startups())
+            self.assign_startups(judge)
+            if judge.remaining <= 0 and not judge.startups:
                 self.judges.remove(judge)
-            self.ticks += 1
-
-    def register_judge_events(self, judge):
-        while judge.events:
-            self.add_event(judge.events.pop(0))
-
-    def assess_allocations(self):
-        for event in self.events:
-            print (event.to_csv())
-        print("done,allocate.py,No more judges,")
-
-    def assess_bins(self):
-        for bin in self.bins:
-            bin.status()
-
-    def add_event(self, event):
-        event.update(time=self.ticks)
-        self.events.append(event)
-
 
-def bin_factory(klass, values, weight):
-    return [klass(value=value, weight=weight)
-            for value in values]
+    def assign_startups(self, judge):
```

Comment:
It depends on the heuristic.  In the case of the RandomSelection and FeatureBins heuristics, both will keep providing applications until the judge capacity is exhausted.  I'm assuming once this gets implemented for real we will just give some them some more applications if they want them.  We'd have to ask a stakeholder if they want to be notified when judges start getting completely satisfied applications.  It's worth noting that this won't be completely obvious from the currently planned analysis interface since there could be a number of queues that still need reads while from particular judges perspective there aren't any applications that they can help satisfy.